ingestion-core/src/main/java/com/mozilla/telemetry/ingestion/core/transform/AddMetadata.java

```diff
@@ -93,43 +126,113 @@ private AddMetadata() {
   }
 
-  private static Map<String, Object> userAgentFromAttributes(Map<String, String> attributes) {
-    HashMap<String, Object> userAgent = new HashMap<>();
-    attributes.keySet().stream() //
-        .filter(k -> k.startsWith(USER_AGENT_PREFIX)) //
-        .forEach(k -> userAgent.put(k.substring(11), attributes.get(k)));
+  static void putIspAttributes(Map<String, String> attributes, ObjectNode metadata) {
+    putAttributes(attributes, metadata, ISP, ISP_PREFIX);
+  }
+
+  static ObjectNode userAgentFromAttributes(Map<String, String> attributes) {
+    ObjectNode userAgent = Json.createObjectNode();
+    attributes.entrySet().stream() //
+        .filter(entry -> entry.getKey().startsWith(USER_AGENT_PREFIX)) //
+        .forEach(entry -> userAgent.put(entry.getKey().substring(USER_AGENT_PREFIX.length()),
+            entry.getValue()));
     return userAgent;
   }
 
-  private static Map<String, Object> headersFromAttributes(Map<String, String> attributes) {
-    HashMap<String, Object> header = new HashMap<>();
-    HEADER_ATTRIBUTES.stream().forEach(name -> Optional //
-        .ofNullable(attributes.get(name)) //
-        .ifPresent(value -> header.put(name, value)));
+  static void putUserAgentAttributes(Map<String, String> attributes, ObjectNode metadata) {
+    putAttributes(attributes, metadata, Attribute.USER_AGENT, USER_AGENT_PREFIX);
+  }
+
+  static ObjectNode headersFromAttributes(Map<String, String> attributes) {
+    ObjectNode header = Json.createObjectNode();
+    attributes.entrySet().stream().filter(entry -> HEADER_ATTRIBUTES.contains(entry.getKey())) //
+        .forEach(entry -> header.put(entry.getKey(), entry.getValue()));
     return header;
   }
 
-  private static Map<String, Object> uriFromAttributes(Map<String, String> attributes) {
-    HashMap<String, Object> uri = new HashMap<>();
-    URI_ATTRIBUTES.stream().forEach(name -> Optional //
-        .ofNullable(attributes.get(name)) //
-        .ifPresent(value -> uri.put(name, value)));
+  static void putHeaderAttributes(Map<String, String> attributes, ObjectNode metadata) {
+    putAttributes(attributes, metadata, HEADER, "");
+  }
+
+  static ObjectNode uriFromAttributes(Map<String, String> attributes) {
+    ObjectNode uri = Json.createObjectNode();
+    attributes.entrySet().stream().filter(entry -> URI_ATTRIBUTES.contains(entry.getKey())) //
+        .forEach(entry -> uri.put(entry.getKey(), entry.getValue()));
     return uri;
   }
+
+  static void putUriAttributes(Map<String, String> attributes, ObjectNode metadata) {
+    putAttributes(attributes, metadata, Attribute.URI, "");
+  }
+
+  ////////
+
+  private static void putAttributes(Map<String, String> attributes, ObjectNode metadata,
+      String nestingKey, String prefix) {
+    Optional //
+        .ofNullable(metadata) //
+        .map(m -> m.path(nestingKey).fields()) //
+        .map(Streams::stream).orElseGet(Stream::empty) //
+        .filter(entry -> !entry.getValue().isNull()) //
```

Comment:
this was modified from `.filter(entry -> entry.getValue() != null) //` because `Json.createObject().set("value", (JsonNode) null).get("value")` returns a `NullNode`, so we should never get a literal `null`.